Show anchor for editing configuration within error text

sagerb opened the pull request:
<!-- Provide a general summary of your changes in the title. -->
<!-- Examples: Updates pull request template -->

## Intent

<!-- Describe what problem you are addressing in this pull request. -->
<!-- If this change is associated with an open issue, please link to it here. -->
<!-- Example: "Resolves #24" -->
<!-- See https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue -->

Resolves #2047 

## Type of Change

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
<!-- If you check more than one box, you may need to refactor this change into separate pull requests -->

- - [ ] Bug Fix <!-- A change which fixes an existing issue -->
- - [x] New Feature <!-- A change which adds additional functionality -->
- - [ ] Breaking Change <!-- A breaking change which causes existing functionality to change -->
- - [ ] Documentation <!-- User or developer documentation -->
- - [ ] Refactor <!-- Code restructuring -->
- - [ ] Tooling <!-- Build, CI, or release scripts and configuration -->

## Approach

<!-- Describe how you solved this problem and any trade-offs you encountered. -->
<!-- Link any additional documentation associated with this change here -->

Implementation looks for a very specific string and if found within the error text, splits the string and inserts an anchor into the HTML. It is only implemented for this single error, but we can move forward with a more generalized approach using the same techniques when we determine other messages that should share this behavior.

![anchor](https://github.com/user-attachments/assets/f02cffab-4022-4541-8180-c9c345e8ecf9)

## Automated Tests

<!-- Describe the automated tests associated with this change. -->
<!-- If automated tests are not included in this change, please state why. -->

## Directions for Reviewers

<!-- Provide steps for reviewers to validate this change manually. -->

Recreate two errors, one of which is the Python mismatch for which this PR is targeting. Confirm this behavior only applies to the Python version mismatch and does not have any impact on other messages.

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply: -->
<!--- If you need clarification on any of these, feel free to ask. We're here to help! -->

- [ ] I have updated [CHANGELOG.md](../CHANGELOG.md) to cover notable changes.

dotNomad reviewed (commented):
[#2119](https://github.com/posit-dev/publisher/issues/2119) is keeping from me from being able to test this without fixes.

When I deploy with an error the deployment is removed from the UI entirely.

Once that is fixed we can come back to this.

sagerb commented:
@dotNomad I think you're good to re-review this. I've merged in Mike's fix and I don't have any problem seeing/viewing deployments with errors now. Can you verify that you are also not blocked?

dotNomad commented on extensions/vscode/webviews/homeView/src/components/EvenEasierDeploy.vue:
```diff
@@ -190,6 +207,9 @@ import { filterConfigurationsToValidAndType } from "../../../../src/utils/filter
 const home = useHomeStore();
 const hostConduit = useHostConduitService();
 
+const editingSplitStr = "editing";
+const editConfigFileDetectionStr = "editing your configuration file";
```
Honestly I would combine these. I think it would be nice if "editing your configuration file" was the link, and it would consolidate a bit of the logic.

sagerb replied:
I did start with that, but didn't like it (see below). The logic consolidation is pretty minor, so I would like to stick with what I've got (the second option below):

![2024-08-22 at 8 57 AM](https://github.com/user-attachments/assets/66ba6a57-3a75-481d-b48e-553e60d02b26)

vs.

![2024-08-22 at 9 08 AM](https://github.com/user-attachments/assets/7a54ee5b-21b7-4a63-b73b-19c702eb3fe4)

dotNomad reviewed (approved):
I was able to see it this time! Works great. Small note about combining some logic.

jonkeane commented:
Sorry I'm a bit late on this: does it look like there's an extra space at the front of the first line?

Also what about "...consider editing your configuration to have one of the versions: 3.11.3"

Slightly more concise + direct. Also would be super slick to condition on number of versions and do "To be version: 3.11.3" if there's one and then "one of versions: ..." if there's more than one

sagerb commented:
> Sorry I'm a bit late on this: does it look like there's an extra space at the front of the first line?
> 
> Also what about "...consider editing your configuration to have one of the versions: 3.11.3"
> 
> Slightly more concise + direct. Also would be super slick to condition on number of versions and do "To be version: 3.11.3" if there's one and then "one of versions: ..." if there's more than one

Space has been eliminated from DOM with the refactoring I just checked in.

We're receiving the error messages from the agent, but I'll go ahead and include the change per your request.

sagerb commented:
@dotNomad Do you mind taking a look at the commit I added to create the new Vue component? https://github.com/posit-dev/publisher/pull/2134/commits/439446684df1887c162f832b219d130022758181

I'll also let you know when I've got the agent code change pushed up.

dotNomad commented on extensions/vscode/webviews/homeView/src/components/TextStringWithAnchor.vue:
```diff
@@ -0,0 +1,63 @@
+  click: [splitOptionId: number]; // named tuple syntax
+}>();
+
+interface SplitOption {
```
May want to export this since it is used as prop, will make it easier to construct with type safety in other components.

sagerb replied:
That has now been refactored into a common utility (and exported), since my next to last commit (counting this one which I'm making your requested changes on) needed the functionality within the log view.

dotNomad commented on extensions/vscode/webviews/homeView/src/components/TextStringWithAnchor.vue:
Splitting this off into a separate component was the right move 🎉

sagerb replied:
I think so too! :-)

dotNomad commented on extensions/vscode/webviews/homeView/src/components/EvenEasierDeploy.vue:
```diff
@@ -349,6 +358,36 @@ const toolTipText = computed(() => {
 - Server URL: ${home.serverCredential?.url || "<undefined>"}`;
 });
 
+const errorMessage = computed(() => {
+  return home.selectedContentRecord?.deploymentError?.msg || "";
+});
```
I would just use `home.selectedContentRecord.deploymentError.msg` directly in the template to get the type narrowing. That way we won't need to have the `|| ""`

sagerb replied:
great! Although I'm not sure why that matches the prop type of string... But it does work. I suppose it is because it still resolves to "" when that message is somehow undefined?

dotNomad reviewed (commented):
Had some comments with the component work.

sagerb commented:
Updated images for latest changes:
![2024-08-23 at 10 34 AM](https://github.com/user-attachments/assets/ba33802f-1231-4333-869b-9fd7d7d17e3f)

![2024-08-23 at 10 35 AM](https://github.com/user-attachments/assets/d96d686a-bf7b-4f3d-9d0b-6944fa423e60)

With these changes, we now display a button within the error message display (at time of publishing) as well as the already implemented link within the last deployment display.

These latest changes also refactored the logic into sharable component and files so that I could implement this new button within the logs view.

@dotNomad Do you mind taking a new look at this PR? Things have dramatically changed since your earlier review.

sagerb commented:
@jonkeane How do the screenshots look to you above? They've got the changes you requested, and I think it was a great improvement.

sagerb commented on extensions/vscode/webviews/homeView/src/components/TextStringWithAnchor.vue:
```diff
@@ -0,0 +1,63 @@
+interface Props {
+  message: string;
+  splitOptions: SplitOption[];
+  class?: string;
```
Since I have multiple root elements within the component, I was doing this explicitly... but looks like I can use `$attrs.class` to go ahead and address that issue in a better way. OK, I will remove that prop.

sagerb commented on extensions/vscode/webviews/homeView/src/components/TextStringWithAnchor.vue:
```diff
@@ -0,0 +1,63 @@
+<script setup lang="ts">
+import { computed } from "vue";
+
+defineOptions({
+  inheritAttrs: false,
+});
```
Good catch, that was left over from a debug session. Sorry about that!

dotNomad commented on extensions/vscode/src/constants.ts:
```diff
@@ -74,6 +74,7 @@ const homeViewCommands = {
   NavigateToDeploymentContent:
     "posit.publisher.homeView.navigateToDeployment.Content",
   ShowContentLogs: "posit.publisher.homeView.navigateToDeployment.ContentLog",
+  EditCurrentConfiguration: "posit.publisher.homeView.edit.Configuration",
```
It looks like the command added was `posit.publisher.edit.Configuration` and did not include the `homeView`. We can change it here or in the `package.json`

dotNomad reviewed (changes requested):
Just caught one thing we have to fix before merging.

dotNomad reviewed (approved):
Thanks for making those strings match :shipit:
